### PR TITLE
Merge openyurt into main

### PR DESCRIPTION
**Pull request by jchua99**
## Summary

3.2.1 To set up Kubernetes and Kubernetes cluster
3.2.2 To set up Knative 
3.2.2 To set up OpenYurt 

Exact commands to run can be found in the readme.

## Implementation Notes :hammer_and_pick:

* The script will assist in deploying kubernetes, and setting up kubernetes cluster in a command, skipping the need to copy paste the token from the master to the slave. 
* The script will deployed necessary knative deployments to support knative function. This is similar to the current installation scripts available. 
* The script will also help to deploy openyurt and its dependencies.

## External Dependencies :four_leaf_clover:

* 

## Breaking API Changes :warning:

* 

*Simply specify none (N/A) if not applicable.*

### CODE REVIEW

**Review by lrq619 — COMMENTED** [review]
Looks fine to me
The runners are down currently, when the runners are all up and all tests passed, this can be merged

File: scripts/openyurt-deployer/README.md
Comment (ustiugov) [inline scripts/openyurt-deployer/README.md:74]: format this as a note so it stands out from the rest of the instructions

File: scripts/openyurt-deployer/README.md
Comment (ustiugov) [inline scripts/openyurt-deployer/README.md:93]:
```suggestion
./openyurt_deployer demo-deploy-on-cloud
```

File: scripts/openyurt-deployer/README.md
Comment (ustiugov) [inline scripts/openyurt-deployer/README.md:97]:
```suggestion
./openyurt_deployer demo-deploy-on-edge
```

File: scripts/openyurt-deployer/main.go
Comment (ustiugov) [inline scripts/openyurt-deployer/main.go:95]: are these modes supported? the readme mentions them

**Review by ustiugov — CHANGES_REQUESTED** [review]
I skimmed the files, the code quality is good and naming is good. The architecture makes sense. Good job, Jason and Ruiqi!

- [ ] move yaml files to `scripts/openyurt-deployer/configs` folder to avoid cluttering of the enclosing folder
- [ ] check that copyrights were added to all the files

File: scripts/openyurt-deployer/main.go
Comment (jchua99) [reply scripts/openyurt-deployer/main.go:95]: yes, edited again in readme

File: scripts/openyurt-deployer/configs/kubeTemplate.yaml
Comment (ustiugov) [inline scripts/openyurt-deployer/configs/kubeTemplate.yaml:4]: what is this address:port?

File: scripts/openyurt-deployer/main_test.go
Comment (ustiugov) [inline scripts/openyurt-deployer/main_test.go:178]: there should not be any stranded commented blocks of code
either remove or restructure the code, or create an issue about why it is commented out, leaving a comment in the code (here)

File: scripts/openyurt-deployer/node.go
Comment (ustiugov) [inline scripts/openyurt-deployer/node.go:50]: clean up

File: scripts/openyurt-deployer/node.go
Comment (ustiugov) [inline scripts/openyurt-deployer/node.go:81]: clean up

File: scripts/openyurt-deployer/node.go
Comment (ustiugov) [inline scripts/openyurt-deployer/node.go:1]: this file has many long lines of code. Please break the long lines into multiple

File: .github/workflows/openyurt-unit-test.yml
Comment (ustiugov) [inline .github/workflows/openyurt-unit-test.yml:41]: very opaque name... what are we testing?

Comment (ustiugov) [inline .github/workflows/openyurt-unit-test.yml:41]: combine all the setup steps in a single step where you execute a single setup script. Run the actual test(s) in a separate step

File: .github/workflows/openyurt-unit-test.yml
Comment (ustiugov) [inline .github/workflows/openyurt-unit-test.yml:117]: same here

**Review by ustiugov — CHANGES_REQUESTED** [review]
- [ ] ask Ruiqi for a detailed code review
- [ ] add the info about OpenYurt in the main README and leave pointers to the supplementary readme files)
- [x] check that the linter checks in the CI are configured and work properly (make sure all the files are linted). you can ask Ruiqi for the details
- [x] make sure that all new .md files are checked for grammar in the CI 
- [x] the output of the added tests is very obscure: it is not clear what is tested and the output is quite verbose with a lot of unnecessary info. Please clean up
- [x] remove `openyurt` branch from the CI workflow configs (openyurt's and other tests will run when opening a PR to `main`)
- [ ] other minor comments

File: scripts/openyurt-deployer/configs/kubeTemplate.yaml
Comment (jchua99) [reply scripts/openyurt-deployer/configs/kubeTemplate.yaml:4]: i have double check this. this is the default template in openyurt to reconfigure the kubelet

**Review by lrq619 — DISMISSED** [review]
I have made a detailed code review, I didn't see any problem.

File: scripts/openyurt-deployer/unit_test_workflow.sh
Comment (ustiugov) [inline scripts/openyurt-deployer/unit_test_workflow.sh:28]: it would be better to separate the actual test from the setup script into a separate CI workflow step. Then, we would be able to clearly distinguish when setup fails and when the test fails

File: scripts/openyurt-deployer/main_test.go
Comment (ustiugov) [inline scripts/openyurt-deployer/main_test.go:116]: These values have to be updated upon every version upgrade. I don't think it's convenient to have them hardcoded. we can have a separate config file (or better: use the versions from the existing config files)

File: scripts/openyurt-deployer/main_test.go
Comment (ustiugov) [inline scripts/openyurt-deployer/main_test.go:105]: remove or uncomment

File: scripts/openyurt-deployer/node.go
Comment (ustiugov) [inline scripts/openyurt-deployer/node.go:668]: remove or uncomment

**Review by ustiugov — CHANGES_REQUESTED** [review]
This part is not done:
> add the info about OpenYurt in the main README and leave pointers to the supplementary readme files)

- [ ] People won't be able to find out that vHive supports OpenYurt if there is nothing in the main README of this repo. 
- [ ] Detail the PR description appropriately

File: scripts/openyurt-deployer/unit_test_workflow.sh
Comment (jchua99) [reply scripts/openyurt-deployer/unit_test_workflow.sh]: i have tried this numerous time. it does not work out as the ssh permission set up in previous steps would not persist to the next step. 

level=error msg="failed to create network operation not permitted:"

File: README.md
Comment (ustiugov) [inline README.md:34]: there is no link to the OpenYurt readme. please add. Specifying the script name in the main readme is too much detail.

**Review by yulinzou — CHANGES_REQUESTED** [review]
Encountered problem while running the demo
`[15:51:52] [Success] Start to init demo
[15:51:52] [Info] Creating benchmark's yaml file and apply it >>>>> 
[15:51:52] [Warn] node: [yulin001@128.110.96.59] failed to exec: 
yq '.metadata.name = \"helloworld-python-cloud\" | 
			.spec.template.spec.nodeSelector.\"apps.openyurt.io/nodepool\" = cloud | 
			.spec.template.spec.containers[0].image = \"docker.io/vhiveease/hello-cloud:latest\" ' 
			configs/benchmarkTemplate.yaml > cloud-bench.yaml
error:bash: line 3: yq: command not found
bash: line 4: configs/benchmarkTemplate.yaml: No such file or directory
[15:51:52] [Error] Process exited with status 127
[15:51:52] [Error] benchmark command fail.
[15:51:52] [Warn] node: [yulin001@128.110.96.59] failed to exec: 
kubectl apply -f cloud-bench.yaml
error: error: no objects passed to apply
[15:51:52] [Error] Process exited with status 1
[15:51:52] [Error] Failed to create benchmark's yaml file and apply it
[15:51:52] [Success] Demo finished!`

**Comment by ustiugov** [comment]
@jchua99 remove the merge commit. you need to use rebase instead.

**Comment by jchua99** [comment]
image

for ./openyurt_deployer demo-deploy-on-cloud

image

for ./openyurt_deploy demo

**Review by yulinzou — DISMISSED** [review]

File: README.md
Comment (ustiugov) [inline README.md:34]:
```suggestion
vHive has added support for the state-of-the-art extension [eStargz](https://github.com/containerd/stargz-snapshotter) to container layers and lazy pull support for container images.

vHive supports running Knative on top of [`OpenYurt`](https://openyurt.io/), an open platform that extends upstream Kubernetes to run on edge node pools. More details on how to the Knative-atop-OpenYurt mode are described [here](scripts/openyurt-deployer/README.md).
```

please fix the above line about stargz in a similar way.

File: scripts/openyurt-deployer/README.md
Comment (ustiugov) [inline scripts/openyurt-deployer/README.md:5]: this is irrelevant. the code in vHive doesn't use any of the old code (I hope)

File: scripts/openyurt-deployer/README.md
Comment (ustiugov) [inline scripts/openyurt-deployer/README.md:2]: add an overview of the Knative-atop-OpenYurt mode. You can give the architecture overview figure (from our tutorial slides) and a pointer to the tutorial slides.

**Review by ustiugov — CHANGES_REQUESTED** [review]
- [x] Please iterate on the readme
- [ ] add a line in the changelog file

File: README.md
Comment (jchua99) [reply README.md]: added line to changelog under unreleased

**Review by ustiugov — COMMENTED** [review]
I am ok. Wait for Yulin's approval and merge.

**Review by ustiugov — DISMISSED** [review]

File: scripts/openyurt-deployer/README.md
Comment (yulinzou) [inline scripts/openyurt-deployer/README.md:9]:
```suggestion
OpenYurt extends Kubernetes from a single cluster to a federation of clusters, offers a seamless way to manage distributed nodes and applications efficiently. This guide outlines how to establish a Kubernetes cluster via kubeadm and then deploy `OpenYurt` and Knative on it, which is compatible with vHive stock-only mode.
```

File: scripts/openyurt-deployer/README.md
Comment (yulinzou) [inline scripts/openyurt-deployer/README.md:16]:
```suggestion
`OpenYurt` is designed to ensure that applications running at the edge of the network remain operational, even in poor connectivity conditions. `OpenYurt` addresses this issue by implementing a per-node proxy (`YurtHub`) along with local storage to cache the state of the cloud apiserver. Consequently, when a node loses its connection, the cached states remain accessible to Kubelet, `KubeProxy`, and any user Pods.
```

File: scripts/openyurt-deployer/README.md
Comment (yulinzou) [inline scripts/openyurt-deployer/README.md:20]:
```suggestion
In an edge computing Kubernetes cluster, nodes are often distributed across various geographical regions. The edge nodes fell within same physical region are referred to as `Pools` in OpenYurt. Consequently, when relying on a native Container Network Interface (CNI) solution, Pods within different `NodePools` may be unable to communicate using Pod IP, Service IP, or Node IP, particularly if each `NodePool` resides within its own isolated LAN. OpenYurt introduces a gateway called `Raven`, which offers a networking solution that enables `cross-NodePool` communication within an `OpenYurt` cluster.
```

File: scripts/openyurt-deployer/README.md
Comment (yulinzou) [inline scripts/openyurt-deployer/README.md:24]:
```suggestion
In order to manage applications and traffic in multiple node pools conveniently, `YurtAppSet` and `YurtAppDaemon` are introduced for managing workloads in `multi-nodepool`, and service topology capability (geographical location, network latency, etc.) for routing traffic in node pool.
```

**Review by yulinzou — APPROVED** [review]

**Review by lrq619 — APPROVED** [review]

**Pull request merged**
